src/transformers/models/emformer/modeling_emformer.py

```diff
@@ -0,0 +1,1158 @@
+        batch_size = utterance.size(1)
+        time_length = right_context.size(0) + utterance.size(0) + summary.size(0)
+
+        # Compute query with [right context, utterance, summary].
+        query = self.emb_to_query(torch.cat([right_context, utterance, summary]))
+
+        # Compute key and value with [mems, right context, utterance].
+        key, value = self.emb_to_key_value(torch.cat([mems, right_context, utterance])).chunk(chunks=2, dim=2)
+
+        if left_context_key is not None and left_context_val is not None:
+            right_context_blocks_length = time_length - torch.max(lengths).int() - summary.size(0)
+            key = torch.cat(
+                [
+                    key[: mems.size(0) + right_context_blocks_length],
+                    left_context_key,
+                    key[mems.size(0) + right_context_blocks_length :],
+                ],
+            )
+            value = torch.cat(
+                [
+                    value[: mems.size(0) + right_context_blocks_length],
+                    left_context_val,
+                    value[mems.size(0) + right_context_blocks_length :],
+                ],
+            )
+
+        # Compute attention weights from query, key, and value.
+        reshaped_query, reshaped_key, reshaped_value = [
+            tensor.contiguous().view(-1, batch_size * self.num_heads, self.input_dim // self.num_heads).transpose(0, 1)
+            for tensor in [query, key, value]
+        ]
+        attention_weights = torch.bmm(reshaped_query * self.scaling, reshaped_key.transpose(1, 2))
+
+        # Compute padding mask.
+        padding_mask = _gen_padding_mask(utterance, right_context, summary, lengths, mems, left_context_key)
+
+        # Compute attention probabilities.
+        attention_probs = self._gen_attention_probs(attention_weights, attention_mask, padding_mask)
+        if output_attentions:
+            attention_outputs = attention_probs.reshape(batch_size, self.num_heads, time_length, time_length)
+            utterance_start = right_context.size(0)
+            utterance_end = right_context.size(0) + utterance.size(0)
+            attention_outputs = attention_outputs[:, :, utterance_start:utterance_end, utterance_start:utterance_end]
+        else:
+            attention_outputs = None
+
+        # Compute attention.
+        attention = torch.bmm(attention_probs, reshaped_value)
+        # attention.shape == (batch_size * self.num_heads, time_length, self.input_dim // self.num_heads)
+        attention = attention.transpose(0, 1).contiguous().view(time_length, batch_size, self.input_dim)
+
+        # Apply output projection.
+        output_right_context_mems = self.out_proj(attention)
+
+        summary_length = summary.size(0)
+        output_right_context = output_right_context_mems[: time_length - summary_length]
+        output_mems = output_right_context_mems[time_length - summary_length :]
+        if self.tanh_on_mem:
```

Comment:
are both branches needed or is one branch also True? Would try to remove as many if-else statements as possible if all checkpoints only ever use one branch